lua/firvish/keymap.lua

```diff
@@ -0,0 +1,231 @@
+-- NOTE: (mostly) copied from astronauta.nvim
+-- See: https://github.com/neovim/neovim/pull/13823
+local keymap = {}
+
+-- Have to use a global to handle re-requiring this file and losing all of the keymap.
+--  In the future, the C code will handle this.
+__FirvishMapStore = __FirvishMapStore or {}
+keymap._store = __FirvishMapStore
+
+keymap._create = function(f)
+    table.insert(keymap._store, f)
+    return #keymap._store
+end
+
+keymap._execute = function(id)
+    return keymap._store[id]()
+end
+
+keymap._expr = function(id)
+    return vim.api.nvim_replace_termcodes(keymap._store[id](), true, true, true)
+end
+
+local make_mapper = function(mode, defaults, opts)
+    local args, map_args = {}, {}
+    for k, v in pairs(opts) do
+        if type(k) == "number" then
+            args[k] = v
+        else
+            map_args[k] = v
+        end
+    end
+
+    local lhs = opts.lhs or args[1]
+    local rhs = opts.rhs or args[2]
+    local map_opts = vim.tbl_extend("force", defaults, map_args)
+
+    local mapping
+    if type(rhs) == "string" then
+        mapping = rhs
+    elseif type(rhs) == "function" then
+        local func_id = keymap._create(rhs)
+
+        if map_opts.expr then
+            mapping = string.format([[luaeval("require("firvish.keymap")._expr(%s)")]], func_id)
+        else
+            assert(map_opts.noremap, "If `rhs` is a function and it's not an expr, `opts.noremap` must be true")
+            mapping = string.format([[<cmd>lua require("firvish.keymap")._execute(%s)<CR>]], func_id)
+        end
+    else
+        error("Unexpected type for rhs:" .. tostring(rhs))
+    end
+
+    if not map_opts.buffer then
+        vim.api.nvim_set_keymap(mode, lhs, mapping, map_opts)
+    else
+        -- Clear the buffer after saving it
+        local buffer = map_opts.buffer
+        if buffer == true then
+            buffer = 0
+        end
+
+        map_opts.buffer = nil
+
+        vim.api.nvim_buf_set_keymap(buffer, mode, lhs, mapping, map_opts)
+    end
+end
+
+--- Helper function for ':map'.
+---
+--@see |vim.keymap.nmap|
+---
```

Comment:
The two ways I know of doing it are (1) [the way neovim does it](https://github.com/neovim/neovim/blob/master/scripts/gen_vimdoc.py) and (2) using TJ's [tree-sitter-lua docgen module](https://github.com/tjdevries/tree-sitter-lua#how-to-generate-documentation). I haven't actually tried. For this PR I'll just move the documentation to the vimdoc. Later, maybe, I'll take a stab at generating the docs.

Reply:
I thought I deleted this comment before submitting the review. You don't need to bother with this since it's a copy of another plug-in. I had a different understanding of it when I wrote the comment.